src/librustc_data_structures/profiling.rs

```diff
@@ -585,3 +601,82 @@ fn get_resident() -> Option<usize> {
         }
     }
 }
+
+fn pass_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| profiler.alloc_string(llvm_name))
+}
+
+fn ir_name_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    llvm_name: &str,
+) -> StringId {
+    *string_cache.entry(llvm_name.to_string()).or_insert_with(|| {
+        let demangled_ir_name = rustc_demangle::demangle(llvm_name).to_string();
+        profiler.alloc_string(demangled_ir_name.as_str())
+    })
+}
+
+fn llvm_args_to_string_id(
+    profiler: &Arc<SelfProfiler>,
+    string_cache: &mut FxHashMap<String, StringId>,
+    pass_name: &str,
+    ir_name: &str,
+) -> EventId {
+    let pass_name = pass_name_to_string_id(profiler, string_cache, pass_name);
+    let mut components = vec![StringComponent::Ref(pass_name)];
+    // handle that LazyCallGraph::SCC is a comma separated list within parentheses
+    let parentheses: &[_] = &['(', ')'];
+    let trimed = ir_name.trim_matches(parentheses);
+    for part in trimed.split(',') {
+        let ir_name = ir_name_to_string_id(profiler, string_cache, part);
+        components.push(StringComponent::Value(SEPARATOR_BYTE));
+        components.push(StringComponent::Ref(ir_name));
+    }
+    EventId::from_label(profiler.alloc_string(components.as_slice()))
+}
+
+pub struct LlvmSelfProfiler<'a> {
+    profiler: Arc<SelfProfiler>,
+    stack: Vec<TimingGuard<'a>>,
+    string_cache: FxHashMap<String, StringId>,
```

Comment:
I have a PR in the works that allows for caching non-`'static` strings in the `SelfProfiler` so this separate cache won't be needed soon.

Reply:
that sounds great. this was kind of a hack to avoiding the file to get larger than 1GB

Reply:
have remove the string_cache and uses the new get_or_alloc_cached_string one down side with the change is that I need to demangle all the time but before I only did it when the string was cached.